Asynchronous Requests Throw Exceptions

Not sure if metrics were made for this, but they throw an exception whenever code within a new timer block starts running while code within an existing timer block is still running.

This issue is specific to mango, where the server handles different http requests, and each http request handler is wrapped in a timer. Due to requests coming in at different times, a timer block will be triggered while another timer block is already being run. 

Ideally we'd like to be able to instrument these asynchronous requests. 

The exception and stack trace is shown below:
```
2016-03-01 15:37:21 WARN  ServletHandler:590 - Error for /reads/chr
java.lang.AssertionError: assertion failed: Timer name from on top of stack [/GET variants(3,false)] did not match passed-in timer name [GET alignment]
	at scala.Predef$.assert(Predef.scala:179)
	at org.bdgenomics.utils.instrumentation.MetricsRecorder.finishPhase(MetricsRecorder.scala:55)
```

--- a/utils-metrics/src/main/scala/org/bdgenomics/utils/instrumentation/MetricsRecorder.scala
+++ b/utils-metrics/src/main/scala/org/bdgenomics/utils/instrumentation/MetricsRecorder.scala
@@ -44,16 +44,32 @@ class MetricsRecorder(val accumulable: Accumulable[ServoTimers, RecordedTiming],
 
   def startPhase(timerName: String, sequenceId: Option[Int] = None, isRDDOperation: Boolean = false) {
     val newSequenceId = generateSequenceId(sequenceId, timerName)
-    val shouldRecord = if (timingsStack.isEmpty) true else shouldRecordOperation(isRDDOperation, timingsStack.top)
-    val key = new TimingPathKey(timerName, newSequenceId, isRDDOperation, shouldRecord)
-    val newPath = if (timingsStack.isEmpty) root(key) else timingsStack.top.child(key)
-    timingsStack.push(newPath)
+    this.synchronized {
+      val shouldRecord = if (timingsStack.isEmpty) true else shouldRecordOperation(isRDDOperation, timingsStack.top)
+      val key = new TimingPathKey(timerName, newSequenceId, isRDDOperation, shouldRecord)
+      val newPath = if (timingsStack.isEmpty) root(key) else timingsStack.top.child(key)
+      timingsStack.push(newPath)
+    }
   }
 
   def finishPhase(timerName: String, timingNanos: Long) {
-    val top = timingsStack.pop()
-    assert(top.timerName == timerName, "Timer name from on top of stack [" + top +
-      "] did not match passed-in timer name [" + timerName + "]")
+    var top: TimingPath = null
+    this.synchronized {
+      top = timingsStack.pop()
+      // Stack is out of order. This should never happen for thread-local contexts.
+      if (top.timerName != timerName) {
+        val tmpStack = new mutable.Stack[TimingPath]()
+        while (top.timerName != timerName && !timingsStack.isEmpty) {
+          tmpStack.push(top)
+          top = timingsStack.pop()
+        }
+        assert(top.timerName == timerName, "Timer name from on top of stack [" + top +
+          "] did not match passed-in timer name [" + timerName + "]")
+        while (!tmpStack.isEmpty) {
+          timingsStack.push(tmpStack.pop())
+        }
+      }
+    }
     if (top.shouldRecord) {
       accumulable += new RecordedTiming(timingNanos, top)
     }

